upload/src/addons/SV/WarningImprovements/XF/Service/User/Warn.php

```diff
@@ -1,12 +1,80 @@
 <?php
 
 namespace SV\WarningImprovements\XF\Service\User;
+use XF\Entity\WarningDefinition;
 
 /**
  * Extends \XF\Service\User\Warn
  */
 class Warn extends XFCP_Warn
 {
+    public function setFromDefinition(WarningDefinition $definition, $points = null, $expiry = null)
+    {
+        if ($definition->warning_definition_id !== 0)
+        {
+            $customWarningDefinition = $this->getCustomWarningDefinition();
+
+            if ($points === null)
+            {
+                $points = $customWarningDefinition->points_default;
+            }
+
+            if ($expiry === null)
+            {
+                if ($customWarningDefinition->expiry_type == 'never')
+                {
+                    $expiry = 0;
+                }
+                else
+                {
+                    $expiry = strtotime('+' . $customWarningDefinition->expiry_default . ' ' . $customWarningDefinition->expiry_type);
+                }
+            }
+
+            // if the expiry is too far in the future, just make it actually be permanent
+            if ($expiry >= pow(2, 32) - 1)
+            {
+                $expiry = 0;
+            }
+        }
+
+        /** @var \SV\WarningImprovements\XF\Entity\WarningDefinition $definition */
+        $return = parent::setFromDefinition($definition, $points, $expiry);
+
+        if ($definition->warning_definition_id === 0)
+        {
+            $this->warning->hydrateRelation('Definition', $definition);
+        }
+
+        if ($definition->sv_custom_title || $definition->warning_definition_id === 0)
+        {
+            if (!empty(\SV\WarningImprovements\Listener::$warningInput))
```

Comment:
Redundant check with `empty(\SV\WarningImprovements\Listener::$warningInput['custom_title'])` as the next line.

Reply:
Done